Refactor syntax unicode xinit

livecodefraser opened the pull request.

Commit:
The X_init/X_open process on OSX now supports Unicode

The UTF-8 argv/envp strings are now converted to StringRefs and processed in that form.

Windows and Linux have not been updated, however.

Commit:
Add StringRef X_init support for Win32

Requires testing!

Commit:
Add StringRef X_init support for Linux

Requires testing!

Commit:
Update calls to setlocale()

Calls to the C library's setlocale function have been changed to more appropriate values. Previously, they were setting the LC_CTYPE and LC_COLLATE to the default system locales (as specified, e.g, through the $LANG or $LC_ALL environment variables) when what they really wanted to do was set them to the US English settings with the native encoding (Codepage 1252 for Windows, ISO-8859-1 for Linux and MacRoman for OSX). The other locale settings are set to the system default (this includes C library error messages, numerical formatting and time/dat formatting).

Commit:
Win32 elevated slaves now pass argv/envp as UTF-16

Passing the string arrays in this format allows Unicode strings to be passed along to the elevated process.

Commit:
Add MCStringCreateWithSysString and MCStringConvertToSysString

These functions are used on Linux to convert between ISO-8859-1 or UTF-16 encoded StringRefs used by LiveCode and whatever string encoding the Linux system expects (probably UTF-8 or ISO-8859-1 but it could be anything). Windows doesn't need this as its *W() functions always take UTF-16 and on OSX the BSD components should always use UTF-8.

Currently, these functions are only used for ARGV/ENVP strings but ideally should be used whenever a string is passed to a system interface (file APIs, for example).

String handling when spawning elevated processes is not yet correct - it currently assumes that the system encoding is UTF-8 when passing argument strings. This affects both the exec of gksu and the creation of the final elevated process.

Commit:
Process launching on OSX now encodes arguments in UTF-8

livecodesebastien commented on engine/src/dsklnxmain.cpp:
```diff
@@ -73,3 +115,126 @@ int main(int argc, char *argv[], char *envp[])
+		else
+		{
+			// No error, conversion should be complete.
+			assert(in_len == 0);
```
Use `MCAssert()` instead of assert, as it won't compile on Linux.

Commit:
Replace an assert with MCAssert

Commit:
Merge branch 'refactor-syntax_unicode' into refactor-syntax_unicode_xinit

Conflicts:
	engine/src/dskw32.cpp
	engine/src/lnxelevate.cpp
	engine/src/lnxspec.cpp
	engine/src/w32spec.cpp

Commit:
Merge branch 'refactor-syntax_unicode_xinit2' into refactor-syntax_unicode_xinit3

Conflicts:
	engine/src/dskw32.cpp
	engine/src/lnxelevate.cpp
	engine/src/lnxspec.cpp
	engine/src/w32spec.cpp

Commit:
Merge branch 'refactor-syntax_unicode_xinit3' into refactor-syntax_unicode_xinit

Commit:
Merge branch 'refactor-syntax_unicode' into refactor-syntax_unicode_xinit

Commit:
Merge branch 'refactor-syntax_unicode' into refactor-syntax_unicode_xinit

Commit:
Merge branch 'refactor-syntax_unicode' into refactor-syntax_unicode_xinit

Commit:
Merge branch 'refactor-syntax_unicode_xinit' of https://github.com/runrevfraser/livecode into refactor-syntax_unicode_xinit

Commit:
Merge branch 'refactor-syntax_unicode_xinit' of https://github.com/runrevfraser/livecode into refactor-syntax_unicode_xinit

Commit:
Merge branch 'refactor-syntax_unicode_xinit' of https://github.com/runrevfraser/livecode into refactor-syntax_unicode_xinit

Conflicts:
	engine/src/dsklnxmain.cpp

Commit:
Add byte order mark (BOM) support when creating StringRefs with UTF-16

The UTF-16 -> native conversion code should not attempt to convert the BOM to a native char.

Commit:
ifg

Commit:
Merge branch 'refactor-syntax_unicode_xinit' of https://github.com/runrevfraser/livecode into refactor-syntax_unicode_xinit

Commit:
Don't attempt to build dsklnx.cpp on OSX

Commit:
Merge branch 'refactor-syntax_unicode_xinit' of https://github.com/runrevfraser/livecode into refactor-syntax_unicode_xinit

livecodesebastien commented:
The branch compiles and works fine on Linux, Windows and Mac, but still doesn't compile on Android and iOS (errors in function `X_init()` from `mblmain.cpp`)

Commit:
Unicode X_init support for iOS

Commit:
StringRef X_init support for Android

Commit:
Remove an accidentally-committed debug modification

livecodesebastien commented:
Can you merge refactor-syntax_unicode in?

Commit:
Merge branch 'refactor-syntax_unicode' into refactor-syntax_unicode_xinit

Conflicts:
	engine/src/w32spec.cpp

livecodesebastien commented:
There is a problem with this branch, it won't compile on Windows. After the first merging of refactor-syntax_unicode in, errors have been introduced in `dskw32.cpp`, where `doc` should have been updated to `p_doc` like it is in `refactor-syntax_unicode` - I guess this happened because you updated the format string and the signature has been changed in `unicode`.
Also
- it's now broken on iOS - `X_init()` returns false
- it doesn't compile anymore on Android - missing a file entry for `exec-strings-chunk.cpp` in `engine/Android.mk`

Commit:
Fix bug in iOS startup routine

livecodesebastien commented:
Could you merge in refactor-syntax_unicode? This will allow the compilation process to be checked for Android, as for now it fails because of a missing file engine/Android.mk

Commit:
Merge remote-tracking branch 'upstream/refactor-syntax_unicode' into refactor-syntax_unicode_xinit

livecodesebastien commented:
The branch doesn't compile on Windows, you need to update InitializeFiberContext.
Also, for an unknown reason, you have errors on dskw32.cpp - certainly a merge which went silently wrong (see 4 comments above)

Commit:
Merge remote-tracking branch 'upstream/refactor-syntax_unicode' into refactor-syntax_unicode_xinit

Commit:
Fix Win32 compilation problems

Commit:
Merge remote-tracking branch 'upstream/refactor-syntax_unicode' into refactor-syntax_unicode_xinit

Commit:
Merge remote-tracking branch 'upstream/refactor-syntax_unicode' into refactor-syntax_unicode_xinit

livecodesebastien commented:
Still crashes on Windows, `argv` is never initialized in `WinMain` - dskw32main.cpp

Commit:
Correct handling of argv on Win32

runrevmark commented on libfoundation/src/foundation-unicodechars.cpp:
```diff
@@ -225,6 +225,12 @@ bool MCUnicodeCharsMapToNative(const unichar_t *p_uchars, uindex_t p_uchar_count
 
 	for(uindex_t i = 0; i < p_uchar_count; i++)
 	{
+		// Check for a UTF-16 byte order mark (BOM) at the beginning of the string
```
This isn't the right place for handling a BOM - this function is for converting a sequence of UTF-16 codeunits to native, it assumes they are already in the correct byte-order with no BOM... If this function is needed in a context where a BOM might be present, the decoding of the input needs to happen (to the point of getting a correct sequence of UTF-16 codeunits) before this is called. (From the point of view of this function, 0xFEFF and 0xFFFE are perfectly valid - mapping to the 'undefined' char).

livecodesebastien commented on engine/src/dsklnxmain.cpp:
```diff
@@ -73,3 +116,122 @@ int main(int argc, char *argv[], char *envp[])
+{
+	// Create the pseudo-FD that iconv uses for character conversion. The most
+	// convenient form is UTF-16 as StringRefs can be constructed directly from that.
+	iconv_t t_fd = iconv_open("UTF-16", MCsysencoding);
```
You need to specify the endianness here, otherwise `iconv` adds a BOM and all the strings generated have a `?` as first character - and that's why it kept crashing on my machine

Commit:
Get iconv to use non-BOM UTF-16 encodings
